### DISCUSSION TITLE
System.NullReferenceException in CreateOrUpdateBitmap of WritableBitmapRenderHandler

### DISCUSSION THREAD

**Discussion by awbushnell** [cefsharp/CefSharp#4612] [Q&A] [answered]
I am starting this as a discussion as I am new here and not sure how to best get it in as a bug report.   In summary, we have a large C++/C++CLI/C# application that embeds CEFSharp in some of our views.  In addition, we have a rather robust "journaling" system that allows us to playback interactions with the UI and other things on our application..   Recently, we noticed that we are experiencing sporadic NullReferenceExceptions when playing back some of our journals for testing.. I have traced down the issue to the way threading and related optimizations work with WritableBitmapRenderHandler  and specifically the CreateOrUpdateBitmap handler. 

The issue is that the it appears when timing is just right on our playbacks, the following code triggers, with no bitmap having been constructed (I.e. image.Source is null)

https://github.com/cefsharp/CefSharp/blob/2cb81aa943ed3a9787709bfa1764e91abf2bfbe3/CefSharp.Wpf/Rendering/WritableBitmapRenderHandler.cs#L158

I hand traced the threading and it seems that what is going is that background things are happening quickly enough such that one of the following "short circuits" break out of the first time delegate is run, leaving no bitmap ever created:

https://github.com/cefsharp/CefSharp/blob/2cb81aa943ed3a9787709bfa1764e91abf2bfbe3/CefSharp.Wpf/Rendering/WritableBitmapRenderHandler.cs#L122

We locally have a fix for this and it simply is updating the following line: 

https://github.com/cefsharp/CefSharp/blob/2cb81aa943ed3a9787709bfa1764e91abf2bfbe3/CefSharp.Wpf/Rendering/WritableBitmapRenderHandler.cs#L137

to read as follows to make sure image.Source is always valid...

`if (createNewBitmap || image.Source is null)`

**Comment by amaitland** [comment, marked as answer]
> to read as follows to make sure image.Source is always valid...
> 
> `if (createNewBitmap || image.Source is null)`

Pull requests are welcome, please create a PR with your proposed changes 👍

**Reply by awbushnell** [reply]
@amaitland https://github.com/cefsharp/CefSharp/pull/4619  thank you!
